src/caffe/layers/softmax_loss_layer.cpp

```diff
@@ -59,16 +67,19 @@ void SoftmaxWithLossLayer<Dtype>::Forward_cpu(
   int count = 0;
   Dtype loss = 0;
   for (int i = 0; i < outer_num_; ++i) {
-    for (int j = 0; j < inner_num_; j++) {
-      const int label_value = static_cast<int>(label[i * inner_num_ + j]);
-      if (has_ignore_label_ && label_value == ignore_label_) {
-        continue;
+    for (int k = 0; k < label_num_; ++k) {
+      for (int j = 0; j < inner_num_; j++) {
+        const int label_value = static_cast<int>(
+            label[(i * label_num_ + k) * inner_num_ + j]);
+        if (has_ignore_label_ && label_value == ignore_label_) {
+          continue;
+        }
+        DCHECK_GE(label_value, 0);
+        DCHECK_LT(label_value, prob_.shape(softmax_axis_));
+        loss -= log(std::max(prob_data[i * dim + label_value * inner_num_ + j],
```

Comment:
Are you sure is this right?
It shouldn't be something like 
```loss -= log(std::max(prob_data[(i * dim) + (dim/label_num_*k) + label_value * inner_num_ + j], Dtype(FLT_MIN)));``` ?
I'm not sure how you think to feed `bottom[0]` to match the dimension of the labels, it shouldn't be larger, with a size of `previous_size*label_num_`?
Let's say we had 3-classes single labels, so an `INNER_PRODUCT` layer with `num_output: 3` was enough. Now if we have for each input 2 labels each one with 3-classes, the `INNER_PRODUCT` should have `num_output: 6` and you should iterate on the `prob_` blob with an offset that considers both the number of the classes and the size of the labels.
Is it right or I'm missing something?